metadata/lib.rs

```diff
@@ -0,0 +1,632 @@
+//!
+//! // First, we need to parse Cargo.toml.
+//! let source_root = env!("CARGO_MANIFEST_DIR");
+//! let metadata = Metadata::from_crate_root(&source_root)?;
+//!
+//! // Next, learn what arguments we need to pass to `cargo`.
+//! let targets = metadata.targets();
+//! let mut cargo_args = metadata.cargo_args();
+//! cargo_args.push(targets.default_target.into());
+//!
+//! // Now, set up the `Command`
+//! let mut cmd = Command::new("cargo");
+//! cmd.args(cargo_args);
+//! for (key, value) in metadata.environment_variables() {
+//!     cmd.env(key, value);
+//! }
+//!
+//! // Finally, run `cargo doc` on the directory.
+//! let result = cmd.output()?;
+//! # Ok(())
+//! # }
+//! ```
+
+use std::collections::{HashMap, HashSet};
+use std::io;
+use std::path::Path;
+
+use serde::Deserialize;
+use thiserror::Error;
+use toml::Value;
+
+/// The target that `metadata` is being built for.
+///
+/// This is directly passed on from the Cargo [`TARGET`] variable.
+///
+/// [`TARGET`]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts
+pub const HOST_TARGET: &str = env!("DOCS_RS_METADATA_HOST_TARGET");
+/// The targets that are built if no `targets` section is specified.
+///
+/// Currently, this is guaranteed to have only [tier one] targets.
+/// However, it may not contain all tier one targets.
+///
+/// [tier one]: https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-1
+pub const DEFAULT_TARGETS: &[&str] = &[
+    "i686-pc-windows-msvc",
+    "i686-unknown-linux-gnu",
+    "x86_64-apple-darwin",
+    "x86_64-pc-windows-msvc",
+    "x86_64-unknown-linux-gnu",
+];
+
+/// The possible errors for [`Metadata::from_crate_root`].
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum MetadataError {
+    /// The error returned when the manifest could not be read.
+    #[error("failed to read manifest from disk")]
+    IO(#[from] io::Error),
+    /// The error returned when the manifest could not be parsed.
+    #[error("failed to parse manifest")]
+    Parse(#[from] toml::de::Error),
+}
+
+/// Metadata to set for custom builds.
+///
+/// This metadata is read from `[package.metadata.docs.rs]` table in `Cargo.toml`.
+///
+/// An example metadata:
+///
+/// ```text
+/// [package]
+/// name = "test"
+///
+/// [package.metadata.docs.rs]
+/// features = [ "feature1", "feature2" ]
+/// all-features = true
+/// no-default-features = true
+/// default-target = "x86_64-unknown-linux-gnu"
+/// targets = [ "x86_64-apple-darwin", "x86_64-pc-windows-msvc" ]
+/// rustc-args = [ "--example-rustc-arg" ]
+/// rustdoc-args = [ "--example-rustdoc-arg" ]
+/// ```
+///
+/// You can define one or more fields in your `Cargo.toml`.
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct Metadata {
+    /// List of features to pass on to `cargo`.
+    ///
+    /// By default, docs.rs will only build default features.
+    features: Option<Vec<String>>,
```

Comment:
I'd rather this be named `docsrs_features` or `enabled_features` since I'm going to be having this parse all features soon

Reply:
It's named after the format in `Cargo.toml`, which can't be changed without breaking all crates. I can use `serde(rename = "enabled_features")` if you prefer?

Reply:
Does this get the features table of the Cargo.toml or the `package.metadata.docs.rs.features` section?

Reply:
The second, `package.metadata.docs.rs.features`.